File: airbyte-server/src/test/java/io/airbyte/server/apis/PermissionApiControllerTest.java
/*
 * Copyright (c) 2020-2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis;

import io.airbyte.api.model.generated.PermissionCheckRead;
import io.airbyte.api.model.generated.PermissionCheckRequest;
import io.airbyte.api.model.generated.PermissionCreate;
import io.airbyte.api.model.generated.PermissionIdRequestBody;
import io.airbyte.api.model.generated.PermissionRead;
import io.airbyte.api.model.generated.PermissionReadList;
import io.airbyte.api.model.generated.PermissionUpdate;
import io.airbyte.api.model.generated.PermissionsCheckMultipleWorkspacesRequest;
import io.airbyte.api.model.generated.UserIdRequestBody;
import io.airbyte.config.persistence.ConfigNotFoundException;
import io.airbyte.validation.json.JsonValidationException;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpStatus;
import java.io.IOException;
import java.util.UUID;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

@SuppressWarnings("PMD.JUnitTestsShouldIncludeAssert")
class PermissionApiControllerTest extends BaseControllerTest {

  @Test
  void testCreatePermission() throws IOException, JsonValidationException {
    Mockito.when(permissionHandler.createPermission(Mockito.any()))
        .thenReturn(new PermissionRead());
    final String path = "/api/v1/permissions/create";
    testEndpointStatus(
        HttpRequest.POST(path, new PermissionCreate().workspaceId(UUID.randomUUID())),
        HttpStatus.OK);
  }

  @Test
  void testGetPermission() throws ConfigNotFoundException, IOException {
    Mockito.when(permissionHandler.getPermission(Mockito.any()))
        .thenReturn(new PermissionRead());
    final String path = "/api/v1/permissions/get";
    testEndpointStatus(
        HttpRequest.POST(path, new PermissionIdRequestBody()),
        HttpStatus.OK);
  }

  @Test
  void testUpdatePermission() throws ConfigNotFoundException, IOException {
    final UUID userId = UUID.randomUUID();
    Mockito.when(permissionHandler.getPermission(Mockito.any()))
        .thenReturn(new PermissionRead().userId(userId));
    final String path = "/api/v1/permissions/update";
    testEndpointStatus(
        HttpRequest.POST(path, new PermissionUpdate().permissionId(UUID.randomUUID())),
        HttpStatus.OK);
  }

  @Test
  void testDeletePermission() {
    Mockito.doNothing().when(permissionHandler).deletePermission(Mockito.any());
    final String path = "/api/v1/permissions/delete";
    testEndpointStatus(
        HttpRequest.POST(path, new PermissionIdRequestBody()),
        HttpStatus.OK);
  }

  @Test
  void testListPermissionByUser() throws IOException {
    Mockito.when(permissionHandler.listPermissionsByUser(Mockito.any()))
        .thenReturn(new PermissionReadList());
    final String path = "/api/v1/permissions/list_by_user";
    testEndpointStatus(
        HttpRequest.POST(path, new UserIdRequestBody()),
        HttpStatus.OK);
  }

  @Test
  void testCheckPermission() throws IOException {
    Mockito.when(permissionHandler.checkPermissions(Mockito.any()))
        .thenReturn(new PermissionCheckRead());
    final String path = "/api/v1/permissions/check";
    testEndpointStatus(
        HttpRequest.POST(path, new PermissionCheckRequest()),
        HttpStatus.OK);
  }

  @Test
  void testCheckMultipleWorkspacesPermission() throws IOException {
    Mockito.when(permissionHandler.permissionsCheckMultipleWorkspaces(Mockito.any()))
        .thenReturn(new PermissionCheckRead());
    final String path = "/api/v1/permissions/check_multiple_workspaces";
    testEndpointStatus(
        HttpRequest.POST(path, new PermissionsCheckMultipleWorkspacesRequest()),
        HttpStatus.OK);
  }

}

File: airbyte-server/src/test/kotlin/io/airbyte/server/apis/controllers/PermissionApiControllerTest.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis.controllers

import io.airbyte.api.model.generated.PermissionCheckRead
import io.airbyte.api.model.generated.PermissionCheckRequest
import io.airbyte.api.model.generated.PermissionCreate
import io.airbyte.api.model.generated.PermissionIdRequestBody
import io.airbyte.api.model.generated.PermissionRead
import io.airbyte.api.model.generated.PermissionReadList
import io.airbyte.api.model.generated.PermissionType
import io.airbyte.api.model.generated.PermissionUpdate
import io.airbyte.api.model.generated.PermissionsCheckMultipleWorkspacesRequest
import io.airbyte.api.model.generated.UserIdRequestBody
import io.airbyte.commons.server.handlers.PermissionHandler
import io.airbyte.config.Permission
import io.airbyte.server.assertStatus
import io.airbyte.server.status
import io.micronaut.context.ApplicationContext
import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpStatus
import io.micronaut.http.client.HttpClient
import io.micronaut.http.client.annotation.Client
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import io.mockk.every
import io.mockk.mockk
import jakarta.inject.Inject
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import java.util.UUID

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@MicronautTest
internal class PermissionApiControllerTest {
  @Inject
  lateinit var context: ApplicationContext

  lateinit var permissionHandler: PermissionHandler

  @Inject
  @Client("/")
  lateinit var client: HttpClient

  @BeforeAll
  fun setupMock() {
    permissionHandler = mockk()
    context.registerSingleton(PermissionHandler::class.java, permissionHandler)
  }

  @Test
  fun testCreatePermission() {
    every { permissionHandler.createPermission(any()) } returns
      Permission()
        .withPermissionType(Permission.PermissionType.WORKSPACE_ADMIN)

    val path = "/api/v1/permissions/create"
    assertStatus(
      HttpStatus.OK,
      client.status(
        HttpRequest.POST(
          path,
          PermissionCreate()
            .workspaceId(UUID.randomUUID())
            .permissionType(PermissionType.WORKSPACE_ADMIN),
        ),
      ),
    )
  }

  @Test
  fun testGetPermission() {
    every { permissionHandler.getPermissionRead(any()) } returns PermissionRead()

    val path = "/api/v1/permissions/get"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, PermissionIdRequestBody())))
  }

  @Test
  fun testUpdatePermission() {
    val userId = UUID.randomUUID()
    every { permissionHandler.getPermissionRead(any()) } returns PermissionRead().userId(userId)
    every { permissionHandler.updatePermission(any()) } returns Unit

    val path = "/api/v1/permissions/update"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, PermissionUpdate().permissionId(UUID.randomUUID()))))
  }

  @Test
  fun testDeletePermission() {
    every { permissionHandler.deletePermission(any()) } returns Unit

    val path = "/api/v1/permissions/delete"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, PermissionIdRequestBody())))
  }

  @Test
  fun testListPermissionByUser() {
    every { permissionHandler.permissionReadListForUser(any()) } returns PermissionReadList()

    val path = "/api/v1/permissions/list_by_user"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, UserIdRequestBody().userId(UUID.randomUUID()))))
  }

  @Test
  fun testCheckPermission() {
    every { permissionHandler.checkPermissions(any()) } returns PermissionCheckRead()

    val path = "/api/v1/permissions/check"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, PermissionCheckRequest())))
  }

  @Test
  fun testCheckMultipleWorkspacesPermission() {
    every { permissionHandler.permissionsCheckMultipleWorkspaces(any()) } returns PermissionCheckRead()

    val path = "/api/v1/permissions/check_multiple_workspaces"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, PermissionsCheckMultipleWorkspacesRequest())))
  }
}
